src/ring_buffer.c

```diff
@@ -0,0 +1,83 @@
+/**
+ * @Author: S. Sharma <silentcat>
+ * @Date:   2019-03-07T10:32:56-06:00
+ * @Email:  silentcat@protonmail.com
+ * @Last modified by:   silentcat
+ * @Last modified time: 2019-03-07T22:58:26-06:00
+ */
+
+#include "include/ring_buffer.h"
+#include <stdio.h>
+
+enum cc_stat rbuf_new(Rbuf **rbuf)
+{
+    RbufConf conf;
+    rbuf_conf_init(&conf);
+    return rbuf_conf_new(&conf, rbuf);
+}
+
+enum cc_stat rbuf_conf_new(RbufConf *rconf, Rbuf **rbuf)
+{
+    Rbuf *ringbuf = rconf->mem_alloc(sizeof(Rbuf));
+    if (!ringbuf)
+       return CC_ERR_ALLOC;
+    if (!(ringbuf->buf = rconf->mem_calloc(DEFAULT_RBUF_CAPACITY, sizeof(Rbuf)))) {
+      rconf->mem_free(ringbuf);
+      return CC_ERR_ALLOC;
+    }
+
+    for (int i = 0; i < DEFAULT_RBUF_CAPACITY; i++)
+        ringbuf->buf[i] = 0x0;
+
+    ringbuf->mem_alloc = rconf->mem_alloc;
+    ringbuf->mem_calloc = rconf->mem_calloc;
+    ringbuf->mem_free = rconf->mem_free;
+    ringbuf->capacity = DEFAULT_RBUF_CAPACITY;
+    ringbuf->size = 0;
+    ringbuf->head = 0, ringbuf->tail = 0;
+    *rbuf = ringbuf;
+    return CC_OK;
+}
+
+enum cc_stat rbuf_conf_init(RbufConf *rconf)
+{
+    rconf->capacity = DEFAULT_RBUF_CAPACITY;
+    rconf->mem_alloc = malloc;
+    rconf->mem_calloc = calloc;
+    rconf->mem_free = free;
+    return CC_OK;
+}
+
+void rbuf_destroy(Rbuf *rbuf)
+{
+    rbuf->mem_free(rbuf->buf);
+    rbuf->mem_free(rbuf);
+}
+
+int rbuf_is_empty(Rbuf *rbuf)
+{
+    return (rbuf->size == 0);
+}
+
+enum cc_stat rbuf_enqueue(Rbuf *rbuf, uint64_t item)
+{
+    if (!rbuf)
```

Comment:
It doesn't make much sense to return `CC_ERR_ALLOC` if there's no allocation going on. 

Also, the null check isn't much of a protection either. An invalid pointer will rarely be `0`. If someone passes a `free`d pointer this check will still pass. If someone does this:
```
Rbuf *rb;
rbuf_enqueue(rb, 42);
```
the pointer will be uninitialized, but still not `0`. It'll be some random stack garbage, which will make the test pass yet again.

It makes sense to check for `0` when it actually means something, like `malloc` returning `0` to indicate failure. It's useless for validating pointers.